File: Stacks-And-Queues/Challenge-3.js
function createSetStack(capacity) {
  const stacks = [[]];

  function push(value) {
    const currentStack = stacks[stacks.length - 1];

    if (currentStack.length < capacity) {
      currentStack.push(value);
    } else {
      stacks.push([value]);
    }
  }
  function pop() {
    const currentStack = stacks[stacks.length - 1];

    if (currentStack.length === 0 && stacks.length > 1) {
      stacks.pop();
    }
    if (stacks.length > 0) {
      return stacks[stacks.length - 1].pop();
    } else {
      return undefined;
    }
  }
  function display() {
    console.log(stacks);
  }

  return { push, pop, display };
}

const setOfStacks = createSetStack(3);

setOfStacks.push('3');
setOfStacks.push('2');
setOfStacks.push('8');
setOfStacks.push('1');
setOfStacks.push('4');

console.log(setOfStacks.display());

class CreateSetStack {
  constructor(capacity, stacks = [[]]) {
    this.capacity = capacity;
    this.stack = stacks;
  }
  push(value) {
    const currentStack = this.stack[this.stack.length - 1];

    if (currentStack.length < this.capacity) {
      currentStack.push(value);
    } else {
      this.stack.push([value]);
    }
  }
}

File: Stacks-And-Queues/Challenge-3.py
class SetOfStacks:
  def __init__(self,capacity):
    self.capacity = capacity
    self.stack = []
    self.current_stack = []
  def push(self,item):
    if len(self.current_stack) == self.capacity:
      self.stack.append(self.current_stack)
      self.current_stack = [item]
    else:
      self.current_stack.append(item)
  def pop(self):
    if not self.stack and not self.current_stack:
      raise IndexError("pop from an empty SetOfStacks")

    if not self.current_stack:
      self.current_stack = self.stack.pop()

    return self.current_stack.pop()
  def show_stack(self):
    print('stack', self.stack, 'current_stack', self.current_stack)


set_of_stacks = SetOfStacks(2)

set_of_stacks.push(1)
set_of_stacks.push(2)
set_of_stacks.push(5)
set_of_stacks.push(8)
set_of_stacks.push(3)
set_of_stacks.push(12)
set_of_stacks.show_stack()
